apps/recommendation/api/src/huggy/core/model_engine/factory.py

```diff
@@ -0,0 +1,90 @@
+from typing import Optional
+
+import huggy.schemas.offer as o
+from huggy.core.model_engine import ModelEngine
+from huggy.core.model_engine.recommendation import Recommendation
+from huggy.core.model_engine.similar_offer import SimilarOffer
+from huggy.schemas.playlist_params import PlaylistParams
+from huggy.schemas.user import UserContext
+from sqlalchemy.ext.asyncio import AsyncSession
+
+
+class ModelEngineOut:
+    model: ModelEngine
+    results: list[str]
+
+
+class ModelEngineFactory:
+    """
+    Factory for creating the appropriate model engine handler.
+    """
+
+    @staticmethod
+    async def handle_prediction(
+        db: AsyncSession,
+        user: UserContext,
+        params_in: PlaylistParams,
+        call_id: str,
+        context: str,
+        *,
+        use_fallback: bool,
+        input_offers: Optional[list[o.Offer]] = None,
+    ) -> ModelEngineOut:
+        """
+        Returns the appropriate model engine based on input context and offers.
+        Fallback to default recommendation if no results are found or specific conditions apply.
+        """
+        if input_offers is None:
+            input_offers = []
+
+        model_engine = None
+
+        # Determine which model to use based on offers and context
+        if any(offer.is_sensitive for offer in input_offers):
+            model_engine = Recommendation(
+                user=user,
+                params_in=params_in,
+                call_id=call_id,
+                context="recommendation",
+                input_offers=input_offers,
+            )
+        elif input_offers and context == "recommendation":
+            model_engine = SimilarOffer(
+                user=user,
+                params_in=params_in,
+                call_id=call_id,
+                context="hybrid_recommendation",
+                input_offers=input_offers,
+            )
+        elif input_offers:
+            model_engine = SimilarOffer(
+                user=user,
+                params_in=params_in,
+                call_id=call_id,
+                context="similar_offer",
+                input_offers=input_offers,
+            )
+        else:
```

Comment:
TODO here we should have recommendation_fallback in case previous context was similar_offer but we did not found any of theses offers within our database